modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.metadata;
+
+import static java.lang.String.format;
+import static org.mule.runtime.core.config.i18n.MessageFactory.createStaticMessage;
+import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.checkInstantiable;
+import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.getAnnotatedFields;
+import org.mule.runtime.api.metadata.MetadataKey;
+import org.mule.runtime.api.metadata.MetadataResolvingException;
+import org.mule.runtime.api.metadata.resolving.FailureCode;
+import org.mule.runtime.api.metadata.resolving.MetadataContentResolver;
+import org.mule.runtime.api.metadata.resolving.MetadataOutputResolver;
+import org.mule.runtime.core.api.MuleRuntimeException;
+import org.mule.runtime.extension.api.annotation.metadata.MetadataKeyId;
+import org.mule.runtime.extension.api.annotation.metadata.MetadataKeyPart;
+import org.mule.runtime.extension.api.introspection.ComponentModel;
+import org.mule.runtime.module.extension.internal.model.property.ImplementingMethodModelProperty;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Parameter;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Instantiates a POJO that is injected in a {@link MetadataContentResolver} or a {@link MetadataOutputResolver}.
```

Comment:
let's make this more Key and Type centered, instead of focusing on java and the resolvers
```Provides an instance of the parameter's type, for a parameter annotated with {@link MetadataKeyId}.
The instance will have populated all the type fields annotated with {@link MetadataKeyPart} with the corresponding value of the represented {@link MetadataKey}```